Move lines up and down with multiple selections

lpommers opened the pull request:
Closes https://github.com/atom/atom/issues/2470

![move-line](https://cloud.githubusercontent.com/assets/5764316/9087105/e7be09fa-3b88-11e5-8059-777b0d902e56.gif)

`ctrl-cmd-up` and `ctrl-cmd-down` now supports moving multiple selections up and down.

Commit:
moveLineUp now loops though all selectedBufferRanges in sorted order so as to return immediately if the first selection is at the start row

Commit:
moveLineDown now loops through all selectedBufferRanges in reverse sorted order so none of the lines move if the bottom-most selection is at the last line

Commit:
:white_check_mark: add specs for moveLineUp and moveLine Down

benogle commented:
cc @nathansobo I wonder if we can find the old specs for these methods. They may have been blown away inadvertently.

nathansobo commented:
I think we should probably just roll forward unless you want to go hunting.

nathansobo commented on spec/text-editor-spec.coffee:
```diff
@@ -1986,6 +1986,82 @@ describe "TextEditor", ->
+      describe "when there is only one line selected", ->
+        it "moves the line up by one row", ->
+          editor.setSelectedBufferRange([[3, 2], [3, 9]])
+          expect(editor.getSelectedBufferRange()).toEqual {start: {row: 3, column: 2}, end: {row: 3, column: 9}}
```
In specs we typically use a shorter array-based notation: `[[3, 2], [3, 9]]`

lpommers replied:
@nathansobo I noticed this is as well... but whenever I ran my tests they failed, asking for the specific range objects instead of the array style notation. I didn't know what else to do; but I did notice the array based styled used elsewhere

nathansobo replied:
Wow, that's odd. There may be some method that's not doing a `Range.fromObject` on its arguments and causing issues, but the equality checks should work. Could you try one more time with this particular assertion and if it doesn't work I'll investigate?

lpommers replied:
Feeling a bit like the boy who cried wolf because I didn't have any problems using the array syntax today.. strange. changes made

nathansobo commented on spec/text-editor-spec.coffee:
```diff
@@ -1986,6 +1986,82 @@ describe "TextEditor", ->
+          expect(editor.getSelectedBufferRange()).toEqual {start: {row: 2, column: 2}, end: {row: 2, column: 9}}
+
+      describe "when there is multiple selections", ->
+
```
:fire: extra newline

nathansobo commented on spec/text-editor-spec.coffee:
```diff
@@ -1986,6 +1986,82 @@ describe "TextEditor", ->
+
+          editor.moveLineUp()
+
+          expect(editor.getSelectedBufferRange()).toEqual {start: {row: 2, column: 2}, end: {row: 2, column: 9}}
```
In addition to the range checks, you should also make assertions about the content of these lines.

nathansobo commented on spec/text-editor-spec.coffee:
```diff
@@ -1986,6 +1986,82 @@ describe "TextEditor", ->
+            expect(editor.getSelectedBufferRanges()).toEqual [
+              {start: {row: 0, column: 2}, end: {row: 0, column: 9}},
+              {start: {row: 2, column: 2}, end: {row: 2, column: 9}},
+              {start: {row: 4, column: 2}, end: {row: 4, column: 9}} ]
```
Another stupid code styling tweak: This `]` should be on the next line.

nathansobo commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,126 @@ class TextEditor extends Model
+    newSelectionBufferRanges = []
+    selections = @getSelectedBufferRanges()
+    selections.sort (a, b) ->
+      return a.compare(b)
```
Don't need `return` in CoffeeScript.

nathansobo commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,126 @@ class TextEditor extends Model
+    for selection in selections
+      return if selection.start.row is 0
+      lastRow = @buffer.getLastRow()
+      return if selection.isEmpty() and selection.start.row is lastRow and @buffer.getLastLine() is ''
```
Is there test coverage for this line? Can you explain why it's needed?

nathansobo commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,126 @@ class TextEditor extends Model
-        if endPosition.row is lastRow and endPosition.column > 0 and not @buffer.lineEndingForRow(endPosition.row)
-          lines = "#{lines}\n"
+        for row in rows
+          if fold = @displayBuffer.largestFoldStartingAtBufferRow(row)
```
I'm actually a bit unclear why the existing code needed to talk explicitly about folds... Is it possible to work strictly in terms of screen coordinates and have things work? Meaning if we're moving a selection that starts at screen row 10 up, it should be safe to insert it at screen row 9 and that should automatically account for folds. Am I missing anything with this thinking? Did you try working exclusively in screen coordinates?

lpommers replied:
I didn't really experiment with screen coordinates. I will spend some time playing around and seeing what happens

lpommers replied:
@nathansobo Been playing with screen coordinates. Just want to make sure I have this straight: In order to insert text at a certain screen row, it's position needs to be converted using something like `bufferPositionForScreenPosition` and then inserted using `insert`, correct?

nathansobo replied:
Correct. Are you encountering problems with that approach. Perhaps there's something I didn't consider.

nathansobo commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,126 @@ class TextEditor extends Model
+          @buffer.insert(insertPosition, lines)
 
-      @setSelectedBufferRange(selection.translate([-insertDelta]), preserveFolds: true, autoscroll: true)
+        # Restore folds that existed before the lines were moved
```
Yeah... my intuition tells me we shouldn't need to mess with unfolding and refolding folds here. Help me understand if I'm wrong.

nathansobo commented:
:heart: Thanks for taking a crack at this. I'd like to see if we can simplify the implementation a bit with respect to fold handling, but this is a great start.

Commit:
CR: remove return statement

Commit:
CR: use array style syntax for bufferRanges

lpommers commented:
@nathansobo I will spend some more time digging into the folds and possibly using screen coordinates. I didn't really investigate too deep when examining this as the main culprit that jumped out was that it just didn't use all the selected buffer ranges.

nathansobo commented:
@lpommers Checking in on this... you ready for another review?

lpommers commented:
@nathansobo unfortunately not. I've spent a bit of time playing with using screen coordinates and am still having some trouble making it work with folds. When I come across a fold, instead of putting the row above it, it places it on the same line as the starting fold. Not sure why yet...

nathansobo commented:
Taking a deeper look here... you're definitely right that my initial suggestion of leaning heavily on screen coordinates isn't enough... folds will have to be dealt with explicitly. Which is a bummer.

Commit:
Wrap all line movements in transact

Commit:
Use row-oriented methods

Commit:
Compute insertDelta more simply

Commit:
Don’t move any lines if some lines can’t move

Commit:
Append line ending if it isn’t present

Commit:
Operate on selection objects, not just ranges

Commit:
Reassign selection range immediately when moving lines up

Commit:
Move all lines spanned by selection in a single operation

Commit:
:tulip: spec gardening

Commit:
Handle multiple selections on the same line

nathansobo commented:
Okay, I've made some changes to your original PR for `move-line-up` on this [branch](https://github.com/atom/atom/compare/move-lines-up-and-down-with-multiple-selections). Would you be interested in taking a look at my adjustments and if you're in favor, add them to this PR and apply a similar treatment to `move-lines-down`?

lpommers commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,133 @@ class TextEditor extends Model
-        @foldBufferRow(foldedRow)
+        # Restore folds that existed before the lines were moved
+        for rangeToRefold in rangesToRefold
+          @displayBuffer.createFold(rangeToRefold.start.row, rangeToRefold.end.row)
```
@nathansobo here don't we want to refold at `rangeToRefold.start.row - 1 ` instead? The fold has moved down one row after the insert, so it's start row is now different. Creating the fold on the start row pushes the folding down one row too many. 
![refold](https://cloud.githubusercontent.com/assets/5764316/9220414/c6a68fb2-40e2-11e5-8402-d47b82efe6fa.gif)

nathansobo replied:
@lpommers Definitely looks wrong, but I think we should correct the problem when we're originally pushing the range onto `rangesToRefold`. Would be good to add a spec that reproduces the behavior you've pictured.

nathansobo commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,133 @@ class TextEditor extends Model
-    return if selection.start.row is 0
-    lastRow = @buffer.getLastRow()
-    return if selection.isEmpty() and selection.start.row is lastRow and @buffer.getLastLine() is ''
+    selections = @getSelectionsOrderedByBufferPosition()
```
Sorry, but I realized when I woke up this morning that we probably don't actually have to refer to the selection objects directly... I had changed it so we could reassign the selection ranges one at a time, but that had bad interactions with selection merging, so I went back to your approach of just resetting the selected buffer ranges at the end. So you can probably switch back to your original approach of just working with the sorted ranges.

nathansobo commented:
@lpommers Ping me on this PR when you're ready for another review!

Commit:
revert to using selection ranges instead of selection objects

Commit:
:white_check_mark: moveLineUp dealing with folds

Commit:
change moveLineDown to handle multiple selections, mimic moveLineUp

Commit:
:sunflower: spec for moveLineDown

lpommers commented:
@nathansobo I've implemented the changes into `moveLineDown` and I think this is ready for another review. One thing I just noticed was that if you move a selection down a row so it is one before a fold it unfolds the row. but this isn't called by `moveLineDown`'s `fold.destroy()`. I looked into it and it might be coming from `@displayBuffer.unfoldBufferRow`, but I don't know why, hoping you can give me some insight here.
![unfold](https://cloud.githubusercontent.com/assets/5764316/9288292/fbd8f954-433f-11e5-8a36-3e51d16d4a77.gif)

nathansobo commented:
Sorry for the delay on this... taking a look at your changes from this weekend now.

nathansobo commented:
So, by default creating a selection that crosses a fold destroys the fold. This is so we don't create hidden selections. You can suppress that behavior by passing `preserveFolds: true` as an option to `setSelectedBufferRanges`.

One thing I'm noticing when I do that however is that lines are getting merged into the fold below them when they're moved down. Are you sure you're re-folding the correct rows? I bet what's happening is that the fold is somehow getting created to include the moved line, then you restore the selection and it destroys that line... check this out with folds preserved on creating the selections:

![move-lines-and-folds](https://cloud.githubusercontent.com/assets/1789/9367705/5c80543e-467c-11e5-8e44-38544e92ba70.gif)

Commit:
fix folds accidentally becoming unfolded

lpommers commented:
@nathansobo yep, that was definitely it!

On a somewhat unrelated note, should I make an issue about the documentation? - I didn't know  `setSelectedBufferRanges` could accept `preserveFolds`, the docs only refer to a `reversed` parameter in the `options` object. https://atom.io/docs/api/v1.0.7/TextEditor#instance-setSelectedBufferRanges

nathansobo commented:
Sorry out the docs. Even better, if you want just change them in a PR and I'll merge it.

nathansobo commented:
Once you think this is working correctly from a behavior perspective, I'll do another review of the actual code. Ping me and @izuzak and we'll hammer on it and try to find holes. Great work on this by the way. This problem is way more challenging than I would have thought and really suggests some improvements to our underlying modeling of folds.

Commit:
update typo in comment

Commit:
use Point as insertion and no need to post-fix newline

Commit:
removed unnecessary + operator

Commit:
preserve folds when making new selections

lpommers commented:
@nathansobo and @izuzak I think we can take another look :eyes:  at this again. I think we're getting there. And thanks for all the help with this, it's been a bit challenging for me and you've all been really helpful

Commit:
now possible to always move lines down at the end of the file

lpommers commented:
this PR now somewhat closes https://github.com/atom/atom/issues/7653, since moving down at the end of the file can now continue infinitely, but you will still always see the next screen row as the last 
![lines](https://cloud.githubusercontent.com/assets/5764316/9429372/1af6d254-49ce-11e5-8d44-6e453d174dba.gif)

Commit:
remove spec pertaining to moving down on last line :sunflower:

izuzak commented:
Hey @lpommers :wave: -- thanks so much for working on this! I was just testing this and noticed one strange thing: when you move selections down then the screen scrolls down correctly when the bottom selection reaches the bottom of the screen, but when you move selections up then the screen scrolls up only after the top selection passes over the top of the screen. Here's a GIF which demonstrates this:

![lines](https://cloud.githubusercontent.com/assets/38924/9446250/d0585ea2-4a90-11e5-95bb-f87e550a084d.gif)

Do you see the same behavior?

I'll play with this a bit more and report if I notice anything else :smile:

izuzak commented:
Noticed one other strange thing -- here's a GIF:

![lines2](https://cloud.githubusercontent.com/assets/38924/9446550/9f84e32a-4a92-11e5-80db-2daf2aaa72cd.gif)

I make two selections on separate lines, and then start moving them down. But notice that what's selected changes as I move the selections down (unexpected). When I move the selections up, the selected text doesn't change (expected)

lpommers commented:
Hi @izuzak, I'm able to reproduce both of those. As to your first comment, I'm not sure what's happening there.. gonna look into it. And the second, I only get that behavior when the two lines are right next to each other - I have an idea as to what might be happening.. gonna look into these (and anything else that pops up) sometime this week when I have some free time.

Commit:
bug fix: go through the selections in reverse order

Commit:
scroll to top-most/bottom-most buffer position

Commit:
spec garden for reversing the buffer selections :tulip:

lpommers commented:
@izuzak I have implemented fixes for those two bugs you spotted, but I think there is a better way to solve the second bug (screen not scrolling at the proper cursor) - perhaps @nathansobo might have a suggestion.

izuzak commented:
:zap: @lpommers thanks for the ping :bow:

> but I think there is a better way to solve the second bug (screen not scrolling at the proper cursor) - perhaps @nathansobo might have a suggestion.

@nathansobo Since you've been :hammer: at scrolling recently (and https://github.com/atom/atom/pull/8548 will be merged soon), do you perhaps want to take a quick look at how things are going here before I test things manually again?

nathansobo commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,131 @@ class TextEditor extends Model
+          newSelectionRanges.push(selection.translate([-insertDelta, 0]))
 
-  # Move lines intersecting the most recent selection down by one row in screen
+      @setSelectedBufferRanges(newSelectionRanges)
```
I'd suggest passing `autoscroll: false` here to suppress any autoscroll before scrolling explicitly.

nathansobo replied:
Maybe you want to `preserveFolds` too just for symmetry?

nathansobo commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,131 @@ class TextEditor extends Model
 
-  # Move lines intersecting the most recent selection down by one row in screen
+      @setSelectedBufferRanges(newSelectionRanges)
+      @scrollToBufferPosition([newSelectionRanges[0].start.row + 2, 0])
```
This seems okay to me if autoscrolling didn't work correctly when setting the ranges. Sometimes it's better to just ask for what you want explicitly and I think this case may be complex enough for it to be one of those times.

lpommers replied:
cool. The reason I wasn't sure about going this route is because I couldn't (and still can't) really see why the autoscrolling wasn't working properly. And I wasn't sure if this kind of solution was sort of duct-tapey without addressing the root of the problem.

nathansobo commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,131 @@ class TextEditor extends Model
+          newSelectionRanges.push(selection.translate([insertDelta, 0]))
 
-      @setSelectedBufferRange(selection.translate([insertDelta]), preserveFolds: true, autoscroll: true)
+       @setSelectedBufferRanges(newSelectionRanges, {preserveFolds: true})
```
Same here... suppress autoscroll.

nathansobo commented:
@izuzak aside from a couple minor adjustments, if you can't break this, I'm ready to merge.

Commit:
set autoscroll to false when setting new buffer ranges

izuzak commented:
I'm still seeing some problems when moving multiple selected lines up. Here's a GIF:

![lines3](https://cloud.githubusercontent.com/assets/38924/9610796/fb1b59b8-50db-11e5-9788-5f7abae8499c.gif)

Notice how when I move the selections down -- the same lines are still selected. But when I move them up -- different lines are selected (then numbers in the selections change).

izuzak commented:
I'm also seeing a strange exception being thrown if I select multiple line, and then press and hold `cmd-z` to undo:

![lines4](https://cloud.githubusercontent.com/assets/38924/9610942/ff78901a-50dc-11e5-9320-b67eef2772d8.gif)

The exception is `Uncaught TypeError: Cannot read property 'getScreenPosition' of undefined` and it's thrown from: `/Users/izuzak/github/atom/src/text-editor.coffee:420`. Wasn't able to reproduce the same thing on the master branch.

@lpommers Are you able to reproduce that exception? @nathansobo might also be interested in this as well.

izuzak commented:
I should have probably mentioned that I'm testing these changes on a merge between the latest master and this branch (so that I'm testing what the behavior will be when this is merged).

Commit:
Merge remote-tracking branch 'upstream/master' into move-lines-up-and-down-with-multiple-selections

lpommers commented:
@izuzak I can only reproduce the first of the two issues. I can't seem to reproduce the `Uncaught TypeError` on a merge between master and the branch. I will look into what's going on with the `moveLineUp` functionality

lpommers commented on src/text-editor.coffee:
```diff
@@ -818,114 +818,131 @@ class TextEditor extends Model
+        selection = selections.shift()
+        selectionsToMove = [selection]
+
+        while selection.end.row is selections[0]?.start.row
```
@nathansobo can you explain to me why this check is necessary? this seems to be the cause of @izuzak most recent finding. When selecting two "sibling" lines via the line numbers, this hits true and the first is removed causing the top line not to move upward. I've only been able to reproduce this gif when selecting lines from the line numbers, not from a cursor in the editor or click and drag - here's his gif of the issue. 

https://cloud.githubusercontent.com/assets/38924/9610796/fb1b59b8-50db-11e5-9788-5f7abae8499c.gif

lpommers replied:
I get this now

lpommers commented:
@izuzak I've just encountered this bug, can you see it as well? It only happens when selecting lines via the line number gutter.. 
![line_down](https://cloud.githubusercontent.com/assets/5764316/9627243/ef99cae8-5162-11e5-8c9e-9af0a2181643.gif)

izuzak commented:
@lpommers Yep, seeing that as well:

![lines6](https://cloud.githubusercontent.com/assets/38924/9641679/43bbb078-51b8-11e5-9398-58e310cc9858.gif)

I'm seeing strange behavior with moving a line that is directly above a folded line even on master (not on your branch):

![lines7](https://cloud.githubusercontent.com/assets/38924/9641744/966e8764-51b8-11e5-8695-c460fbb9b765.gif)

Notice how when I'm not moving the whole line -- things work as expected. But if I select the whole line -- then the folded line moves as well. :confused:

nathansobo commented:
This is getting incredibly tricky. I do think we can keep hacking at this and get it right eventually. I'm wondering if some changes to lower level abstractions might make things easier.

* Folds should really be represented as markers that span the elided text. Right now they start at the beginning of the first line that remains visible, which causes us to stray into them too easily. If the marker for the folded text extended from the end of the first folded line to the end of the fold, maybe we could avoid accidentally going into folds.
* I've always wanted to allow folds to be cut/copied and pasted along with text. I'm wondering if such a facility could make this easier. We'd just take the union of all lines spanned by contiguous selections, cut the content with its fold metadata, then paste it one line down.

I feel like my time might be spent fixing some of these lower level issues, but I will merge this if we can get it working sooner.

jeancroy commented:
Bug https://github.com/atom/atom/issues/8023 migth be related to this discussion of special behavior on the line above the fold.

Commit:
bug that causes moveLineUp to move the wrong selections

lpommers commented:
@izuzak So, I think we're slowly getting there... I've been able to fix two of those bugs. 
* w/r/t the bug where moving two lines down (selecting via the gutter) through a fold caused the lines to separate: I've fixed the worst part - but currently the fold gets destroyed - so i need more time with that.
* Moving lines up should no long move the wrong selections
* still unable to reproduce the thrown exception

here's a gif of what I mean about the fold getting destroyed, but the lines moving properly

![fold](https://cloud.githubusercontent.com/assets/5764316/9734884/f2db207a-5636-11e5-98a3-889414732c7c.gif)

lpommers commented:
I need to change the two commits i just made - forgot to take into consideration multiple selections on the same line..

Commit:
fix for "bug that causes moveLineUp to move the wrong selections"

Commit:
fix for "make one selection to move down if the start row matches the next's end row"

lpommers commented:
@izuzak I think we're getting there... some notes on two of the things I've fixed 

* w/r/t the bug where moving two lines down below a fold (selecting via the line gutter), the issue of the lines breaking apart is solved - but the fold gets destroyed - will need more time
* moving lines up should now have the proper selections
* still haven't been able to reproduce the uncaught exception

gif of what I mean about moving lines down through the fold 
![fold](https://cloud.githubusercontent.com/assets/5764316/9735459/4d73cb96-563b-11e5-8b32-211c519f2947.gif)

lpommers commented:
@izuzak  @nathansobo I think we've :hammer: out most of the kinks with this. I'm wondering if it's worth merging this and then making a new issue or PR to handle any specific issues may that pop up as this has become quite large.

izuzak commented:
@lpommers I applaud your efforts here :clap: :clap: -- thanks for sticking with this. :bow:

Just took this for a test drive again and had a hard time finding problems. I noticed two:

1. The exceptions thrown when I hold and press undo, as described here: https://github.com/atom/atom/pull/8232#issuecomment-136799859. You mentioned that you're unable to reproduce these, so I'd love it if @nathansobo could take a look. I'm unable to trigger these on `master` so I suspect there's something on this branch triggering this. Not sure how serious it is or if it's worth blocking this PR on it.

2. There seems to be a few small problems when moving lines around wrapped lines. Here's an example:

  ```
1
2
Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud
3
4
```

  And here's how that looks on my screen wrapped:

  <img width="644" alt="screen shot 2015-09-11 at 18 42 57" src="https://cloud.githubusercontent.com/assets/38924/9820565/f93234d8-58b4-11e5-8d74-2ff523940703.png">

  A few problems I observed:
  * If I select both lines 1 and 2 (by cmd-clicking in the gutter) and then try to move them down -- nothing happens. If I disable soft wrapping -- I can move them down just fine.
  * If I put a cursor on line 1 and another on line 2 (so, similar case as above, but whole lines are not selected), a strange thing happens when I try to move the lines down -- lines 1 and 2 switch places and it's not possible to move them down anymore. Here's a GIF:

    ![p1](https://cloud.githubusercontent.com/assets/38924/9820676/acf1800a-58b5-11e5-8b0b-5b6835997576.gif)

    Even if you just place a single cursor on line 2 and try to move it down over the wrapped line -- it won't work. As far as I can tell, this problem doesn't happen on master.

> I'm wondering if it's worth merging this and then making a new issue or PR to handle any specific issues may that pop up as this has become quite large.

I wouldn't be :-1: on that, but will defer to @nathansobo to make the call on that.

Thanks again :zap:

lpommers commented:
izuzak I should be applauding you guys :bow: for all the help here. It's easy to stick with this, since it's so fun. Without your help, I would have missed a ton of use cases/scenarios; it's something I want to get better at. I'm gonna spend some time this weekend taking another crack at the issues you brought up

abe33 committed:
Merge branch 'master' into pr/8232

abe33 committed:
:bug: Fix moving multiple selections down locked at wrapped line

abe33 committed:
:art: Fix bad indentation

abe33 committed:
:art: Remove unnecessary branching

abe33 committed:
:art: Fix description test

Actually it doesn’t test multiple selections moves.

abe33 committed:
:bug: Fix folds not preserved when moving multiple selections down

When two or more selections span the rows immediately before a fold,
the `did-change` event dispatched on the deletion will trigger a merge
of the selections, which in turn trigger an unfold at the buffer
position of the new selection, which is now the position of the fold.
Consolidating the selections at the begin of the transaction will
prevent the merge and will keep the fold untouched.

abe33 committed:
:art: Improve test descriptions and fix a typo

Commit:
Merge pull request #2 from abe33/cn-move-lines-up-and-down-with-multiple-selections

Move lines up and down with multiple selections

lpommers commented:
@izuzak @nathansobo Thanks to some great work from @abe33 :bow:, the issue with soft wrapped lines is fixed. Maybe we can get some new :eyes: on this since it's been a little while since there was activity, mostly due to me being too busy lately to have any time to focus on this. hopefully that will change soon.

izuzak commented:
:zap: @abe33 and @lpommers :zap:

Just gave this branch a try again and the soft wrap problem does seem to be fixed, as does the exception I reported in https://github.com/atom/atom/pull/8232#issuecomment-136799859. Great work!

If someone else from @atom/feedback would like to give this branch a try as well -- I think that would be swell.

Also /cc @nathansobo for code reviewing the new changes.

nathansobo commented:
Code looks good. If this is behaving correctly, I'd say :ship: at your discretion @izuzak.

izuzak commented:
> Code looks good. If this is behaving correctly, I'd say :ship: at your discretion @izuzak.

Just gave it another go and I did find two differences compared to master. @lpommers @abe33 Do you see the same difference? Sorry for not noticing them when I first tested this!

1. Moving a wrapped line over the top edge doesn't show the full line (the top of the wrapped line is not shown on this branch, but is shown on master).

  Behavior on master (I just move a wrapped line down and back up -- notice how the top of the line is visible when I start moving it back up and over the top edge of the window):
  
  ![bug1_master](https://cloud.githubusercontent.com/assets/38924/10600871/ed654e02-770d-11e5-9d27-6261397f7f15.gif)

  Behavior on this branch (notice how the top of the wrapped line is not visible when I move it over the top edge. Also notice how the screen "jumps" when I start moving the line up -- that doesn't happen on master either):

  ![bug1_branch](https://cloud.githubusercontent.com/assets/38924/10600878/f82a934c-770d-11e5-9dcf-5fcaa2832651.gif)

2. Moving a folded line breaks the folded line.

  Behavior on master (I just fold a line and move it down, then unfold -- works as expected):

  ![bug2_master](https://cloud.githubusercontent.com/assets/38924/10600907/31b6939a-770e-11e5-83d0-da9c80378656.gif)

  Behavior on this branch (notice how the folded line breaks apart when I try to move it down):

  ![bug2_branch](https://cloud.githubusercontent.com/assets/38924/10600913/38b061b2-770e-11e5-8ebb-7519e9e47d9b.gif)

abe33 commented:
@izuzak Thanks for the feedback, I'll try merging back master in the to see if it's related to changes in master or just caused by an issue in our implementation.

lpommers commented:
@izuzak @abe33 I merged back master locally and still seeing the issues you presented here. Gonna try and spend some time with this..

lpommers commented:
@izuzak I seem to be seeing these issues you mentioned on both `master` and this branch now..

izuzak commented:
> @izuzak I seem to be seeing these issues you mentioned on both master and this branch now..

@lpommers That's strange, I'm still seeing the behavior I described in https://github.com/atom/atom/pull/8232#issuecomment-149463992. On master, moving a folded line doesn't break that line on master, and moving wrapped lines above the top edge always shows the cursor. On this branch (with master merged in), a folded line breaks when you try to move it and cursors on wrapped lines are not always visible when you move them over the top edge.

If you're seeing different behavior, can you share some GIF which clearly show that you're using the latest version of master and the two problems reproduced? That might help us figure out why we're not seeing the same behavior.

Thanks! :bow:

lpommers commented:
@izuzak you're correct. I was looking at the wrong window when I reported that... sorry for the confusion, I got mixed up :flushed:

abe33 committed:
Merge branch 'master' of github.com:atom/atom into cn-move-lines-up-and-down-with-multiple-selections

abe33 commented:
@izuzak @lpommers I'm going to take a look at that this evening.

If I understand correctly: 

1. Solution here is to always make the screen row with the cursor visible at the end of the motion
2. When moving a folded line, it also moves all the hidden lines in the fold without unfolding it.

I'll add tests for these two cases as well.

abe33 committed:
:bug: Fix moving single selection intersecting a fold

abe33 committed:
:art: Fix tests descriptions

abe33 committed:
:bug: Fix moving a multiple selection with a fold creating new folds

abe33 committed:
:bug: Fix moving multiple selection with folds creates new folds

abe33 committed:
:fire: Remove test focus

abe33 commented:
Some news, I fixed most issues with moving folded rows. As folds in `rangesToFold` were not destroyed there was a lot of side effects when moving multiple selections containing folds and close enough that one selection was refolding the previously moved fold (but still present in the display buffer as it wasn't destroyed) with wrong coordinates. 

I just have to fix the issue with scrolling when moving a wrapped line and it should be good for another :eyes:.

Commit:
Merge pull request #3 from abe33/cn-move-lines-up-and-down-with-multiple-selections

Move lines up and down with multiple selections

lpommers commented:
@abe33  seeing some interesting behavior w/ folded rows on this branch now... are you seeing this? I've been experiencing some strange behavior on my local build recently, so it might just be my copy at the moment.. 

![fols](https://cloud.githubusercontent.com/assets/5764316/10917898/50f1d28c-8263-11e5-9ced-4b9edf5f9e47.gif)
![fols2](https://cloud.githubusercontent.com/assets/5764316/10917899/50f56f1e-8263-11e5-935f-00eca807b79d.gif)

maxbrunsfeld commented:
FYI, I just merged https://github.com/atom/atom/pull/9104, a small PR that adds autoindent when moving lines. It introduced a conflict with this branch, but it should be pretty easy to resolve.

abe33 commented:
@lpommers Interesting, actually I'm not a heavy user of folds so I'm mostly working in the dark here, relying on tests to make sure I'm not missing anything, maybe we should another grammar that js in tests as make sure moving lines with folds operate smoothly in every cases. And there's so many cases it's hard to get the whole picture without a good set of tests. I think I'll try to add more tests using the cases you described to make sure I didn't missed some crucial factor.

@maxbrunsfeld thanks for the heads up, I'll take some time tomorrow to review the PR with the new changes.

abe33 committed:
Merge back master

abe33 commented:
@lpommers Are you sure you tried with a version of Atom running the branch? I can't reproduce either issues: 
![moving-html-lines-with-fold](https://cloud.githubusercontent.com/assets/247140/10946438/93bc9c5e-8322-11e5-9a9d-fc3b937b0828.gif)

abe33 committed:
:art: Place move lines test with auto indent with other move lines tests

abe33 committed:
:bug: Fix auto indent no longer applied when moving lines

abe33 commented:
@maxbrunsfeld as you had anticipated it was pretty easy to resolve, I took the liberty to moves the autoindent tests with the other line moves tests so that we don't need to look in several place in the future.

As far as I can tell, the only remaining issue is when moving a wrapped line triggers a scroll. I'm looking into it now.

abe33 committed:
:bug: Fix top of a wrapped line not visible when moving it up

abe33 commented:
Alright, I think we're pretty much done with this PR, if someone want to give it a look.

For the fun, here's an example of moving a multiple selection with a fold and autoindent enabled: 
![moving-html-lines-with-fold-and-autoindent](https://cloud.githubusercontent.com/assets/247140/10947582/05350992-8329-11e5-96df-610a02236d2b.gif)

:sparkles:

maxbrunsfeld commented:
👏 Nice!

Commit:
Merge pull request #5 from abe33/cn-move-lines-up-and-down-with-multiple-selections-2

Move lines up and down with multiple selections

lpommers commented:
Tested :eyes:  this locally and things are looking good. haven't been able to notice anything. :bow: :pray: :clap: to @abe33 for the work on this. And thanks to everyone else.

not entirely sure what that build error is referring to though.. not sure what we can do about the license of a package? 
```
Running "check-licenses" task
Found dependencies without permissive licenses:
spdx-expression-parse@1.0.1
  License: (MIT AND CC-BY-3.0)
  Repository: git+https://github.com/kemitchell/spdx-expression-parse.js
The command "script/cibuild" exited with 1.
```

abe33 commented:
Yes, I just noticed that too, @50Wliu had identified the issue coming from atom/legal-eagle not expecting a parenthesis in the license string (https://github.com/atom/legal-eagle/blob/master/src/legal-eagle.coffee#L47-L56).

abe33 committed:
Merge branch 'master' of github.com:atom/atom into cn-move-lines-up-and-down-with-multiple-selections

50Wliu commented:
Just restarted the build since legal-eagle has been updated on master.  Might take a while to complete though since I also restarted 3 others before this one.

50Wliu commented on src/text-editor.coffee:
```diff
@@ -866,116 +866,177 @@ class TextEditor extends Model
         fn(selection, index) for selection, index in @getSelectionsOrderedByBufferPosition()
 
-  # Move lines intersection the most recent selection up by one row in screen
+  # Move lines intersection the most recent selection or multiple selections up by one row in screen
```
:memo: "intersection" -> "intersecting"

abe33 replied:
Thank you for the heads up, I pushed a fix on @lpommers repo, it'll be available when he'll merge the branch.

abe33 committed:
:art: Fix typo in moveLineUp comment

Commit:
Merge pull request #6 from abe33/cn-move-lines-up-and-down-with-multiple-selections

Move lines up and down with multiple selections

abe33 commented:
@izuzak @nathansobo If you guys could give it a quick :eyes: before we get too far from master again. The build failure is due to symbols-view whose tests was broken by language-javascript, as far as I can tell @50Wliu fixed that on master as.

50Wliu commented:
That was actually an accidental commit that I reverted.  The failing symbols-view spec has been super-flaky for a long time now and can be safely ignored: atom/symbols-view#128.

nathansobo commented:
Insanely good job @lpommers, @abe33 and everyone else involved here. As mentioned, I have some ideas for improvements in core that could simplify this implementation, because it shouldn't be this hard. For now though, happy to have the feature.

izuzak commented:
I :clap: you for the work here, @lpommers and @abe33, and also for sticking with this PR for so long. 

:dancers:
